internal/ethapi/transaction_args.go

```diff
@@ -474,25 +484,27 @@ func (args *TransactionArgs) ToMessage(globalGasCap uint64, header *types.Header
 		// ToMessage recurses once to allow ArbOS to intercept the result for all callers
 		// ArbOS uses this to modify globalGasCap so that the cap will ignore this tx's specific L1 data costs
 		core.InterceptRPCGasCap(&globalGasCap, msg, header, state)
-		return args.ToMessage(globalGasCap, header, nil, runMode) // we pass a nil to avoid another recursion
+		err := args.CallDefaults(globalGasCap, baseFee, chainID, gasLimitNotSetByUser)
+		// If we fail to call defaults, we should panic because it's a programming error since we've already called it once
+		if err != nil {
+			panic(fmt.Sprintf("CallDefaults failed: %v", err))
+		}
+		return args.ToMessage(baseFee, globalGasCap, header, nil, runMode, chainID, gasLimitNotSetByUser) // we pass a nil to avoid another recursion
```

Comment:
I'd like to hear more about that.
Also: @Tristan-Wilson

Reply:
Earlier there was no concept of CallDefaults before calling ToMessage, but now there is.
So when we recurse in ToMessage for 

> // Arbitrum: raise the gas cap to ignore L1 costs so that it's compute-only
	if state != nil {

We need to call CallDefaults to set the new globalGasCap (updated by InterceptRPCGasCap) again.
But because the CallDefaults was already called earlier the the args.Gas is already set to a non zero value (even though this value was not given by the user).
So to allow us to update args.Gas even to though it is set, i introduced a new input to function CallDefaults called gasLimitNotSetByUser which will help in determining whether we should be updating args.Gas even if it holds a non zero value (cause this non zero value was not set by the user).